docs/library/rp2.pio.rst

```diff
@@ -0,0 +1,164 @@
+-----------------------
+
+.. function:: asm_pio(*, out_init=None, set_init=None, sideset_init=None, in_shiftdir=0, out_shiftdir=0, autopush=False, autopull=False, push_thresh=32, pull_thresh=32)
+
+    Assemble a PIO program.
+
+    The following parameters control the initial state of the GPIO pins, as one of `PIO.IN_LOW`, `PIO.IN_HIGH`, `PIO.OUT_LOW` or `PIO.OUT_HIGH`. If the program uses more than one pin, provide a tuple, e.g. ``out_init=(PIO.OUT_LOW, PIO.OUT_LOW)``.
+
+    - *out_init* configures the pins used for ``out()`` instructions.
+    - *set_init* configures the pins used for ``set()`` instructions. There can be at most 5.
+    - *sideset_init* configures the pins used side-setting. There can be at most 5.
+
+    The following parameters are used by default, but can be overriden in `state_machine.init()`:
+
+    - *in_shiftdir* is the default direction the ISR will shift, either `PIO.SHIFT_LEFT` or `PIO.SHIFT_RIGHT`.
+    - *out_shiftdir* is the default direction the OSR will shift, either `PIO.SHIFT_LEFT` or `PIO.SHIFT_RIGHT`.
+    - *push_thresh* is the threshold in bits before auto-push or conditional re-pushing is triggered.
+    - *pull_thresh* is the threshold in bits before auto-push or conditional re-pushing is triggered.
+
+    The remaining parameters are:
+
+    - *autopush* configures whether auto-push is enabled.
+    - *autopull* configures whether auto-pull is enabled.
+
+.. function:: asm_pio_encode(instr, sideset_count)
+
+    Assemble a single PIO instruction. You usually want to use `asm_pio()` instead.
+
+    >>> rp2.asm_pio_encode("set(0, 1)", 0)
+    57345
+
+.. class:: PIOASMError
+
+    This exception is raised from `asm_pio()` or `asm_pio_encode()` if there is an error assembling a PIO program.
+
+
+PIO class
+---------
+
+.. class:: PIO(id)
+
+    Gets the PIO instance numbered *id*. The RP2040 has two PIO instances, numbered 0 and 1.
+
+    Raises a ``ValueError`` if any other argument is provided.
+
+.. method:: pio.add_program(program)
+
+    Add the *program* to the instruction memory of this PIO instance.
+
+    The amount of memory available for programs on each PIO instance is limited. If there isn't enough space left in the PIO's program memory, you'll get ``OSError`` 12 ``ENOMEM``.
+
+.. method:: pio.remove_program([program])
+
+    Remove *program* from the instruction memory of this PIO instance.
+
+    If no program is provided, it removes all programs.
+
+    It is not an error to remove a program which has already been removed.
+
+.. method:: pio.state_machine(id, [program, ...])
+
+    Gets the state machine numbered *id*. On the RP2040, each PIO instance has four state machines, numbered 0 to 3.
+
+    Optionally initialize it with a *program*: see `state_machine.init`.
+
+    >>> rp2.PIO(1).state_machine(3)
+    StateMachine(7)
+
+.. method:: pio.irq(handler=None, trigger=IRQ_SM0|IRQ_SM1|IRQ_SM2|IRQ_SM3, hard=False)
+
+    Returns the IRQ object for this PIO instance.
```

Comment:
I wasn't sure where or how to document the PIO/StateMachine IRQ objects, mostly because I don't understand how they work! :sweat_smile: